cpp/src/groupby/sort/group_scan_util.cuh

```diff
@@ -0,0 +1,116 @@
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/utilities/device_atomics.cuh>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_vector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/functional.h>
+#include <thrust/scan.h>
+
+namespace cudf {
+namespace groupby {
+namespace detail {
+template <aggregation::Kind K>
+struct scan_functor {
+  template <typename T>
+  static constexpr bool is_supported()
+  {
+    if (K == aggregation::SUM)
+      return cudf::is_numeric<T>() || cudf::is_duration<T>() || cudf::is_fixed_point<T>();
+    else if (K == aggregation::MIN or K == aggregation::MAX)
+      return cudf::is_fixed_width<T>() and is_relationally_comparable<T, T>();
+    else
+      return false;
+  }
+
+  template <typename T>
+  std::enable_if_t<is_supported<T>(), std::unique_ptr<column>> operator()(
+    column_view const& values,
+    size_type num_groups,
+    rmm::device_vector<cudf::size_type> const& group_labels,
+    rmm::cuda_stream_view stream,
+    rmm::mr::device_memory_resource* mr)
+  {
+    using DeviceType       = device_storage_type_t<T>;
+    using OpType           = cudf::detail::corresponding_operator_t<K>;
+    using ResultType       = cudf::detail::target_type_t<T, K>;
+    using ResultDeviceType = device_storage_type_t<ResultType>;
+
+    auto result_type = is_fixed_point<T>()
+                         ? data_type{type_to_id<ResultType>(), values.type().scale()}
+                         : data_type{type_to_id<ResultType>()};
+
+    std::unique_ptr<column> result =
+      make_fixed_width_column(result_type, values.size(), mask_state::UNALLOCATED, stream, mr);
+
+    if (values.is_empty()) { return result; }
+
+    auto result_table = mutable_table_view({*result});
+    cudf::detail::initialize_with_identity(result_table, {K}, stream);
+
+    auto resultview = mutable_column_device_view::create(result->mutable_view(), stream);
+    auto valuesview = column_device_view::create(values, stream);
+
+    if (values.has_nulls()) {
+      auto input = thrust::make_transform_iterator(
+        make_null_replacement_iterator(*valuesview, OpType::template identity<DeviceType>()),
+        thrust::identity<ResultDeviceType>{});
+      thrust::inclusive_scan_by_key(rmm::exec_policy(stream),
+                                    group_labels.begin(),
+                                    group_labels.end(),
+                                    input,
+                                    resultview->begin<ResultDeviceType>(),
+                                    thrust::equal_to<size_type>{},
+                                    OpType{});
+      result->set_null_mask(cudf::detail::copy_bitmask(values, stream));
+    } else {
+      auto input = thrust::make_transform_iterator(valuesview->begin<DeviceType>(),
+                                                   thrust::identity<ResultDeviceType>{});
```

Comment:
Is this to convert `DeviceType` to `ResultDeviceType`?

Reply:
Yes. inclusive_scan_by_key doesn't automatically convert. If not converted, it causes compilation errors.